Simplify character handling and speed up compile of WGSL parser

darinadler opened the pull request:
#### 0b0ff310ee75769d5f5462a560aac7f7c44287da
<pre>
Simplify character handling and speed up compile of WGSL parser
<a href="https://bugs.webkit.org/show_bug.cgi?id=299674">https://bugs.webkit.org/show_bug.cgi?id=299674</a>
<a href="https://rdar.apple.com/161488196">rdar://161488196</a>

Reviewed by Sam Weinig.

* Source/WebGPU/WGSL/Lexer.cpp:
(WGSL::Lexer&lt;CharacterType&gt;::makeToken): Moved this out of the header.
(WGSL::Lexer&lt;CharacterType&gt;::makeFloatToken): Ditto.
(WGSL::Lexer&lt;CharacterType&gt;::makeIntegerToken): Ditto.
(WGSL::Lexer&lt;CharacterType&gt;::makeIdentifierToken): Ditto.

* Source/WebGPU/WGSL/Lexer.h: Reduced includes.
Added a forward declaration for TokenType.
(WGSL::Lexer::Lexer): Take a span instead of a String.
(WGSL::Lexer::makeToken): Deleted.
(WGSL::Lexer::makeFloatToken): Deleted.
(WGSL::Lexer::makeIntegerToken): Deleted.
(WGSL::Lexer::makeIdentifierToken): Deleted.

* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::parse): Changed the template argument to be a character type
rather than a Lexer classs type. Pass a span into the Lexer constructor.

* Source/WebGPU/WGSL/ParserPrivate.h: Moved the include of &quot;Token.h&quot;
here from Lexer.h.

* Tools/TestWebKitAPI/Tests/WGSL/ConstLiteralTests.cpp:
(parseLCharPrimaryExpression): Pass a span instead of a String.

* Tools/TestWebKitAPI/Tests/WGSL/LexerTests.cpp:
(TestWGSLAPI::TestLexer::TestLexer): Added an include of &quot;Token.h&quot;.

Canonical link: <a href="https://commits.webkit.org/300659@main">https://commits.webkit.org/300659@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d246bd3ac2e2594efbfcc9a521111595e6e916f3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123383 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43098 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33794 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130097 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/75507 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6b4bb785-a36c-4ffb-b195-d5a93898a045) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125260 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43821 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51692 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/93790 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/75507 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e3af2ec5-1835-4816-9c7a-16bd7214918d) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126336 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34918 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110391 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/74417 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/570a0efc-7d0d-4484-ab4b-b82ee8640cf7) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33886 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/28549 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73614 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104632 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28775 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132814 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50333 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38308 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/102280 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50709 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106614 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102133 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25965 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47489 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25712 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/47123 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50188 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55949 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49659 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53009 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51337 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Tools/TestWebKitAPI/Tests/WGSL/ConstLiteralTests.cpp:
```diff
@@ -33,7 +33,7 @@
 {
     WGSL::ShaderModule shaderModule(input, { });
-    WGSL::Lexer<Latin1Character> lexer(input);
+    WGSL::Lexer<Latin1Character> lexer(input.span<Latin1Character>());
```
Either it should just work, or with a deduction guide, you could probably get this to be:

```
WGSL::Lexer lexer(input.span<Latin1Character>());
```

darinadler replied:
Success! I didn’t need to add a deduction guide

weinig commented on Source/WebGPU/WGSL/Parser.cpp:
```diff
@@ -320,10 +319,10 @@ static AST::UnaryOperation toUnaryOperation(const Token& token)
 std::optional<FailedCheck> parse(ShaderModule& shaderModule)
 {
-    Lexer lexer(shaderModule.source());
+    Lexer<CharacterType> lexer(shaderModule.source().span<CharacterType>());
```
Same here.

darinadler committed:
Simplify character handling and speed up compile of WGSL parser
https://bugs.webkit.org/show_bug.cgi?id=299674
rdar://161488196

Reviewed by Sam Weinig.

* Source/WebGPU/WGSL/Lexer.cpp:
(WGSL::Lexer<CharacterType>::makeToken): Moved this out of the header.
(WGSL::Lexer<CharacterType>::makeFloatToken): Ditto.
(WGSL::Lexer<CharacterType>::makeIntegerToken): Ditto.
(WGSL::Lexer<CharacterType>::makeIdentifierToken): Ditto.

* Source/WebGPU/WGSL/Lexer.h: Reduced includes.
Added a forward declaration for TokenType.
(WGSL::Lexer::Lexer): Take a span instead of a String.
(WGSL::Lexer::makeToken): Deleted.
(WGSL::Lexer::makeFloatToken): Deleted.
(WGSL::Lexer::makeIntegerToken): Deleted.
(WGSL::Lexer::makeIdentifierToken): Deleted.

* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::parse): Changed the template argument to be a character type
rather than a Lexer classs type. Pass a span into the Lexer constructor.

* Source/WebGPU/WGSL/ParserPrivate.h: Moved the include of "Token.h"
here from Lexer.h.

* Tools/TestWebKitAPI/Tests/WGSL/ConstLiteralTests.cpp:
(parseLCharPrimaryExpression): Pass a span instead of a String.

* Tools/TestWebKitAPI/Tests/WGSL/LexerTests.cpp:
(TestWGSLAPI::TestLexer::TestLexer): Added an include of "Token.h".

Canonical link: https://commits.webkit.org/300659@main
